packages/lib/src/verify.ts

```diff
@@ -0,0 +1,245 @@
+  const { R, msgHash } = proof;
+  const { T: computedT, U: computedU } = await recoverTUFromProof(R, msgHash);
+  if (!computedT.equals(T) || !computedU.equals(U)) {
+    return false;
+  }
+  console.timeEnd("T and U Verification");
+
+  console.time("Nullifier Verification");
+  const proofNullifierRandomness = BigInt(publicSignals[6]);
+  if (proofNullifierRandomness !== nullifierRandomness) {
+    return false;
+  }
+  console.timeEnd("Nullifier Verification");
+
+  console.time("Fetching Verification Key");
+  const vKey = isNode()
+    ? await getVerificationKeyFromFile(pathToCircuits!)
+    : await getVerificationKeyFromUrl();
+  console.timeEnd("Fetching Verification Key");
+
+  console.time("ZK Proof Verification");
+  const verified = await verifyMembershipZKP(vKey, proof.zkp);
+  console.timeEnd("ZK Proof Verification");
+
+  // snarkjs will not terminate this object automatically
+  // We should do so after all proving/verification is finished for caching purposes
+  // See: https://github.com/iden3/snarkjs/issues/152
+  // @ts-ignore
+  await globalThis.curve_bn128.terminate();
+  console.timeEnd("Membership Proof Verification");
+
+  return verified;
+};
+
+/**
+ * Verifies a batch of ECDSA membership proofs
+ * Must be used with the same list of public keys and nullifier randomness
+ * Based on the Efficient ECDSA formulation: https://personaelabs.org/posts/efficient-ecdsa-1/
+ * Does not check/maintain the list of usedNullifiers, this must be done by the caller
+ * @param proofs - The membership proofs to verify
+ * @param pubKeys - The list of public keys comprising the anonymity set for the proof
+ * @param nullifierRandomness - Optional nullifier randomness used to generate unique nullifiers
+ * @param pathToCircuits - The path to the verification key. Only needed for server side verification
+ * @returns - A boolean indicating whether or not all of the proofs are valid
+ */
+export const batchVerifyMembership = async (
+  proofs: EcdsaMembershipProof[],
+  pubKeys: WeierstrassPoint[],
+  nullifierRandomness: bigint = BigInt(0),
+  pathToCircuits: string | undefined = undefined
+): Promise<boolean> => {
+  if (isNode() && pathToCircuits === undefined) {
+    throw new Error(
+      "Path to circuits must be provided for server side verification!"
+    );
+  }
+
+  console.time("Batch Membership Proof Verification");
+  console.time("Batch Merkle Root Computation");
+  const edwardsPubKeys = pubKeys.map((pubKey) => pubKey.toEdwards());
+  const computedMerkleRoot = await computeMerkleRoot(edwardsPubKeys);
+  console.timeEnd("Batch Merkle Root Computation");
+
+  console.time("Fetching Verification Key");
+  const vKey = isNode()
+    ? await getVerificationKeyFromFile(pathToCircuits!)
+    : await getVerificationKeyFromUrl();
+  console.timeEnd("Fetching Verification Key");
+
+  const verified = await Promise.all(
+    proofs.map(async (proof, i) => {
+      console.time(`Membership Proof Verification: ${i}`);
```

Comment:
Same comment as with batchProve, should we factor out this code into a separate subroutine? And then use in both this function and the above one?

Reply:
Don't think this is necessary for now, see above.